### ISSUE TITLE
Autodiff support in Numba

### ISSUE THREAD

**Issue by adam-urbanczyk** [numba/numba#8565]
## Feature request
It would be great if Numba supported automatic differentiation. Maybe using  [Enzyme](https://github.com/EnzymeAD/Enzyme) would be the easiest way as it operates directly on the IR of LLVM. Another possible source of inspiration could be [Taichi](https://docs.taichi-lang.org/).

**Comment by gmarkall** [comment]
I'd been meaning to look into Enzyme, as it also seems to support CUDA-specific techniques. @adam-urbanczyk have you any thoughts or experiments in the direction of AD support for Numba that you could add here?

**Comment by adam-urbanczyk** [comment]
Nothing too specific. My use case is estimating probability distributions and I'd be already quite happy with jacobians and hessians of the functions from math and being able to differentiate through if statements.

**Comment by gmarkall** [comment]
I think this is an area none of the Numba maintainers have any significant amount of experience with, so it'll be difficult for us to push forward on this without a significant amount of hand-holding, but maybe we can establish some scope here with a bit of further discussion. To kick this off, I have a few questions:

- What is a small example of a bit of code that you'd like to be able to write?
- What would Numba do / generate under the hood - for example what would some expected transformation of this code look like?
- How was that transformation derived?

**Comment by adam-urbanczyk** [comment]
>     * What is a small example of a bit of code that you'd like to be able to write?

Maybe something like this:

```python

@jit
def f(x,a,b):

   return exp(-a*x**2)*erf(b*x)  

df = f.gradient(a=True, b=True)

a = 1
b = 2
x = np.linspace(-1,1)
f_val = f(x,a,b) # return the function value
f_jacobian = df(x,a,b) # return the gradient w.r.t. a and b. Note that f_jacobian is 2D.
```
 
>     * What would Numba do / generate under the hood - for example what would some expected transformation of this code look like?

I'm just a user of AD so cannot say too much here. My understanding is that Enzyme allows to directly differentiate LLVM IR and Numba generates IR at certain point. It should then "easy" to combine those two tools.

Here is an official example on the IR level from their compiler explorer instance:

```
declare dso_local double @__enzyme_autodiff(i8*, double)

define double @square(double %x) {
entry:
  %mul = fmul double %x, %x
  ret double %mul
}

define double @dsquare(double %x) local_unnamed_addr {
entry:
  %call = tail call double @__enzyme_autodiff(i8* bitcast (double (double)* @square to i8*), double %x)
  ret double %call
}
```
The way I understand it Numba already generates something like `@square` and what would be needed is to add the `@dsquare` part and run the Enzyme transformation on this IR. More examples on IR level are here: https://github.com/EnzymeAD/Enzyme/tree/main/enzyme/test/Enzyme/. Probably `ForwardMode` and `ForwardModeVector` are most usefel for the usacase I proposed at the beginning.
 
>     * How was that transformation derived?

That is in my understanding already handled by Enzyme via the `__enyzme_autodiff` and related functions.

**Comment by wsmoses** [comment]
In contrast Enzyme dev here, with little Numba internal knowledge. Regardless happy to help and this has been something that definitely would be really useful to a lot of people.

In particular what you see by forward, etc is that Enzyme supports several different AD algorithms/modes. Forward mode gets the derivative of all results wrt a single input, reverse mode gets the derivative of all inputs wrt an output (gradient), and vector modes thereof change single in either of the above to one’s choice of N. 

I think the first question here is what communication mechanism makes the most sense. The easiest thing to get working is that Enzyme has a LLVM module pass that replaces all AD calls with a call to a newly generated derivative function. Alternatively, there’s also an API that takes an LLVM function and returns a newly generated derivative function. The API level has more flexibility, but requires a bit more setup. 

Regardless either way I imagine the overall pipeline to be Numba generated LLVM, and via a choice of above, during compilation Numba calls Enzyme to generate the derivative functions.

**Comment by ludgerpaehler** [comment]
Fellow Enzyme-dev here who already looked at Numba and is already working on a prototype - would be really happy to link up to get a more complete coverage and move to have what you are envisioning @adam-urbanczyk 

What I think we can consciously move towards is support for 

- Forward-Mode
- Vector Forward-Mode
- Reverse-Mode
- Vector Reverse-Mode

What kind of target platforms are you thinking of? From an Enzyme perspective Threads, & OpenMP are the easiest. CUDA might be possible? We do not support Intel Threading Building Blocks (TBB) at the current moment, but it would be a matter of software engineering understanding for TBB into Enzyme.

@gmarkall do I understand it correctly that Numba does not use CUDA-Streams, or CUDA-Graphs for its CUDA support?

**Comment by ludgerpaehler** [comment]
cc @ lukasheinrich

**Comment by lukasheinrich** [comment]
Hi,

This is a great initiative. as @ludgerpaehler says, there has already been some effort in this direction.

in terms of API, I think it makes sense to stay as close as possible to the JAX

i.e. 

```
def f(x):
  ...
  
 numba.jvp(f...)
 numba.vjp(f,...)
 numba.grad(f,..)
 numba.value_and_grad(f,..)
```

etc

**Comment by lukasheinrich** [comment]
cc @niklasschmitz

**Comment by gmarkall** [comment]
> @gmarkall do I understand it correctly that Numba does not use CUDA-Streams, or CUDA-Graphs for its CUDA support?

Apologies for the delay in my reply here. It doesn't use graphs, but it supports streams: https://numba.readthedocs.io/en/latest/cuda-reference/host.html#stream-management

**Comment by adam-urbanczyk** [comment]
@gmarkall coming back to the original question, is the description good enough? Do you need more details?

@ludgerpaehler can you share the code you are working on? Do you know if enzyme supports higher order derivatives (specifically calculation of hessians)?

**Comment by wsmoses** [comment]
You can indeed use Enzyme to take arbitrary higher order derivatives.

**Comment by gmarkall** [comment]
> @gmarkall coming back to the original question, is the description good enough? Do you need more details?

I guess so - as my experience in this area is extremely limited, I feel that the most I can do is to help facilitate the discussion and experimentation by those more qualified than myself, and that seems to be happening so I'd imagine it is good enough...

Perhaps others who have commented in this thread can suggest if any more detail or clarification is required, if necessary.

**Comment by adam-urbanczyk** [comment]
@ludgerpaehler is numba-enzyme open sourced / do you plan to open source it?

**Comment by wsmoses** [comment]
Enzyme project lead, but without much numba experience. I haven't seen where any of @ludgerpaehler's tests are or its current state, but I think the first part of an upstream Enzyme-Numba would probably be building an llvmlite (https://github.com/numba/llvmlite) with Enzyme or building Enzyme against the LLVM in the corresponding llvmlite. If there's any llvmlite maintainers around for a chat, I'd be happy to help set it up.

**Comment by clemisch** [comment]
> > Hi :) Giving an updated talk next week at EuroAD, will gladly post the slides after the talk
> 
> Slides appear to be here: https://www.autodiff.org/Docs/euroad/25th%20EuroAd%20Workshop%20-%20Ludger%20Paehler%20-%20Numba-Enzyme:%20Differentiable%20JIT%208217%20d%20Python.pdf

The given example of numba vs JAX performance is very poor IMHO due to using loopy code with JAX. I would be interested in real-world examples which fundamentally require loopy (not vectorized) functions.

**Comment by ludgerpaehler** [comment]
@clemisch 

yes, and no I'd argue - the picked example is a micro-micro benchmark to illustrate the different behavior of the two JIT-compilers. It is just focussing on that one isolated behavior not more, not less. It is definitely not indicative of the behavior on real-world applications where many different language features come together. For the paper we will also have real-world benchmarks which come closer to what you are describing, i.e. non-trivial work inside of the loops.

@adam-urbanczyk 

Sorry for the long silence, yes it is definitely to be open-sourced. Currently pushing to finish the last bits of it for an initial alpha-release in the coming weeks, to then take it from there.  As @wsmoses touched upon, a lot of the stuff Enzyme needs is integrated at the llvmlite level so it is a highly open question how/if that can long-term be upstreamed. But just to set expectations, this is an alpha-release, so it's definitely not at the level where that can, or should be considered.

What the alpha-release will be:

* Forward-mode
* Reverse-mode
* Only on CPU for now

In terms of features etc. we will then have to slowly take it from there. Also, please find [here](https://drive.google.com/file/d/1BOCuu6tnbGabgXaojE4PCz8jT9DYQf_N/view?usp=sharing) the updated master-slides for said presentation with a bit more clarity.

**Comment by adam-urbanczyk** [comment]
> Otherwise, what is the benefit of implementing autodiff with numba? Apart from being an interesting academic problem, of course.

Uhm, weird question. The benefit is that users of numba will be able to use autodiff.

**Comment by ErolBa** [comment]
@ludgerpaehler Is this something you're still working on? Would be fantastic to have this feature, even if it's not polished...

**Comment by ErolBa** [comment]
Alright, thanks. I am unfortunately not knowledgable enough to help, but I hope someone will take up the task.

**Comment by wsmoses** [comment]
That sounds like a reasonable plan to me! Let me know how I can be of help.

**Comment by iguinn** [comment]
Is this feature still being worked on?

**Comment by wsmoses** [comment]
To my knowledge no one has started work on this.

If you're interested and want to work on, let me know ! Happy to provide support

**Comment by adam-urbanczyk** [comment]
@wsmoses too bad that the work of @ludgerpaehler was never open sourced. I'm becoming more and more interested in doing something based on llvm ir pulled out of Numba. So a question to you: are there any projects or examples that do that? I.e. is there an example of wrapping enzyme through Python? Would you have any advice regarding going the X (e.g. Numba) -> *.ll -> *.ll with gradients route?

**Comment by raphaelchinchilla** [comment]
I am also interested in that. It feels that it would be the kind of task that Claude Code would excel at. And having an enzyme powered AD for numba, and if it can be also used for cupy would be to have the promise of what jax should actually have been.

**Comment by wsmoses** [comment]
yeah it should work for both cases [and @adam-urbanczyk sorry I missed your earlier message]. And I agree this is something claude code should definitely be able to help with. If one or both of you are interested in helping spearhead, let me know and I can help point you to the right places. Probably come to the next weekly Enzyme developers meeting and we can chat through more there? Info should be on our website, and if you can't find it feel free to shoot me an email

**Comment by ymardoukhi** [comment]
I have started working on this. I have been actively pursuing the numba/enzyme integration, and it is the core of our software at [QruiseML](https://www.qruise.com/products/qruise-ml). The workflow there is Numba -> LLVM IR -> Julia -> Enzyme.jl. The choice of Julia as a simulation backend is mostly due to the SciML ecosystem. Nevertheless, we have internally decided to open-source this layer of our software.

There are some posts on Julia discourse in this regard

https://discourse.julialang.org/t/get-numba-llvm-ir-differentiated-in-julia-using-enzyme-jl/131255
https://discourse.julialang.org/t/auto-differentiation-of-numba-llvm-ir-using-enzyme-jl-linear-algebra-and-opaque-external-functions/137799

Currently, there are some issues regarding the BLAS and opaque functions in the IR.
https://github.com/EnzymeAD/Enzyme.jl/issues/3300

I am pushing to have an initial version of the package ready before JuliaCon 2026. @wsmoses would be great if I could get some help from you and Valentin.

**Comment by wsmoses** [comment]
Good news, thanks to @ymardoukhi's valiant effort we now have an official numba enzyme integration package, available here: https://github.com/EnzymeAD/numba-enzyme

Obviously this is early days so there is a lot to be done (and I'll let @ymardoukhi say more), but for anyone interested I'll point them to the official numba enzyme repo!

**Comment by ymardoukhi** [comment]
As @wsmoses mentioned, this is the early-stage release of the package, and there is a lot to be added, e.g. support for array inputs (should land in the repo soon with little effort), linear algebra support, binary wheels for different platforms, etc. I really hope this will be an indispensable component for the scientific/numerical computing community. I also wanted to acknowledge the invaluable help @wsmoses, @giordano, and @vchuravy gave me.

**Comment by adam-urbanczyk** [comment]
Thanks @ymardoukhi, great work!
